equational_theories/Completeness.lean

```diff
@@ -0,0 +1,220 @@
+-- Statement and proof of the completeness theorem for magmas,
+-- we roughly follow Baader & Nipkow's "Term Rewriting and All That",
+-- and
+import equational_theories.FreeMagma
+import Mathlib.Data.Set.Defs
+
+#check Set
+
+structure MagmaLaw (α : Type) where
+  lhs : FreeMagma α
+  rhs : FreeMagma α
+deriving DecidableEq
+
+local infix:60 " ≃ " => MagmaLaw.mk
+
+def substFreeMagma {α β} (t : FreeMagma α) (σ : α → FreeMagma β) : FreeMagma β :=
+match t with
+| .Leaf a => σ a
+| t₁ ⋆ t₂ => substFreeMagma t₁ σ ⋆ substFreeMagma t₂ σ
+
+infix:66 " ⬝ " => substFreeMagma
+
+#check ((Lf 0) ⬝ (λ i ↦ Lf i))
+
+@[inline, simp]
+def Ctx α := Set (MagmaLaw α)
+
+#print Membership
+
+#print Set.instMembership
+
+-- FIXME: figure out how to remove this.
+instance Ctx.Membership α : Membership (MagmaLaw α) (Ctx α) := ⟨ Set.instMembership.mem ⟩
+
+section DeriveDef
+
+set_option hygiene false
+
+local infix:50 " ⊢ " =>  derive
+
+-- We keep this in type, because we're going to want to gather
+-- the (finite!) set of required axioms later.
+inductive derive {α} : Ctx α → MagmaLaw α → Type :=
+| Ax Γ E (h : E ∈ Γ) : Γ ⊢ E
+| Ref Γ t : Γ ⊢ (t ≃ t)
+| Sym Γ t u : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ t)
+| Trans Γ t u v : Γ ⊢ (t ≃ u) → Γ ⊢ (u ≃ v) → Γ ⊢ (t ≃ v)
+-- This is not as polymorphic as it could be, shouldn't be an issue at the moment
+| Subst Γ t u σ : Γ ⊢ (t ≃ u) → Γ ⊢ (t ⬝ σ ≃ u ⬝ σ)
+| Cong Γ t₁ t₂ u₁ u₂ : Γ ⊢ (t₁ ≃ t₂) → Γ ⊢ (u₁ ≃ u₂) → Γ ⊢ (t₁ ⋆ u₁ ≃ t₂ ⋆ u₂)
+
+end DeriveDef
+
+def satisfiesPhi {α G : Type} [Magma G] (φ : α → G) (E : MagmaLaw α) : Prop :=
+  evalInMagma φ E.lhs = evalInMagma φ E.rhs
+
+def satisfies {α : Type} (G : Type) [Magma G] (E : MagmaLaw α) := ∀ (φ : α → G), satisfiesPhi φ E
+
+def satisfiesSet {α : Type} (G : Type) [Magma G] (Γ : Set (MagmaLaw α)) : Prop :=
+  ∀ E ∈ Γ, satisfies G E
+
+
+def models {α} (Γ : Ctx α) (E : MagmaLaw α) : Prop :=
+  ∀ (G : Type)[Magma G], satisfiesSet G Γ → satisfies G E
+
+infix:50 " ⊧ " => (satisfies)
+
+infix:50 " ⊧ " => (satisfiesSet)
+
+infix:50 " ⊧ " => (models)
+
+infix:50 " ⊢ " => (derive)
+
```

Comment:
Maybe the definitions up to this point can ultimately end up in their own file, called something like `Satisfaction.lean`? I suggest this because I expect lots of other files (like my `Duality.lean` in progress) will need to use these definitions, but may not need everything in `Completeness.lean`.

Reply:
I agree! I'd like to get this merged first, if that's all right, I'm getting nervous about all the rebasing.